core/services/keeper/orm_test.go

```diff
@@ -166,167 +185,157 @@ func TestKeeperDB_EligibleUpkeeps_Shuffle(t *testing.T) {
 
-	blockheight := int64(120)
-	gracePeriod := int64(100)
-
-	registry, _ := cltest.MustInsertKeeperRegistry(t, db, orm, ethKeyStore)
-	upkeep1 := newUpkeep(registry, 0)
-	upkeep1.LastRunBlockHeight = 0
-	upkeep2 := newUpkeep(registry, 1)
-	upkeep2.LastRunBlockHeight = 19
-	upkeep3 := newUpkeep(registry, 2)
-	upkeep3.LastRunBlockHeight = 20
+	registry, _ := cltest.MustInsertKeeperRegistry(t, db, orm, ethKeyStore, 0, 3, 10)
 
-	for _, upkeep := range [3]keeper.UpkeepRegistration{upkeep1, upkeep2, upkeep3} {
-		err := orm.UpsertUpkeep(&upkeep)
-		require.NoError(t, err)
+	for i := 0; i < 10; i++ {
+		cltest.MustInsertUpkeepForRegistry(t, db, config, registry)
 	}
 
-	cltest.AssertCount(t, db, "upkeep_registrations", 3)
+	cltest.AssertCount(t, db, "keeper_registries", 1)
+	cltest.AssertCount(t, db, "upkeep_registrations", 10)
 
-	eligibleUpkeeps, err := orm.EligibleUpkeepsForRegistry(registry.ContractAddress, blockheight, gracePeriod)
-	assert.NoError(t, err)
-	// 2 out of 3 are eligible, check that ids are 0 or 1 but order is shuffled so can not use equals
-	assert.Len(t, eligibleUpkeeps, 2)
-	assert.Less(t, eligibleUpkeeps[0].UpkeepID, int64(2))
-	assert.Less(t, eligibleUpkeeps[1].UpkeepID, int64(2))
+	// 3 keepers 10 block turns should be different every turn
+	h1 := evmtypes.NewHead(big.NewInt(20), utils.NewHash(), utils.NewHash(), 1000, utils.NewBigI(0))
+	putTurnBlockAsParent(&h1, registry)
+	list1, err := orm.EligibleUpkeepsForRegistry(registry.ContractAddress, &h1, 100)
+	require.NoError(t, err)
+
+	h2 := evmtypes.NewHead(big.NewInt(31), utils.NewHash(), utils.NewHash(), 1000, utils.NewBigI(0))
+	putTurnBlockAsParent(&h2, registry)
+	list2, err := orm.EligibleUpkeepsForRegistry(registry.ContractAddress, &h2, 100)
+	require.NoError(t, err)
+
+	h3 := evmtypes.NewHead(big.NewInt(42), utils.NewHash(), utils.NewHash(), 1000, utils.NewBigI(0))
+	putTurnBlockAsParent(&h3, registry)
+	list3, err := orm.EligibleUpkeepsForRegistry(registry.ContractAddress, &h3, 100)
+	require.NoError(t, err)
+
+	h4 := evmtypes.NewHead(big.NewInt(53), utils.NewHash(), utils.NewHash(), 1000, utils.NewBigI(0))
+	putTurnBlockAsParent(&h4, registry)
+	list4, err := orm.EligibleUpkeepsForRegistry(registry.ContractAddress, &h4, 100)
+	require.NoError(t, err)
+
+	// sort before compare
+	sort.Slice(list1, func(i, j int) bool {
+		return list1[i].UpkeepID < list1[j].UpkeepID
+	})
+	sort.Slice(list2, func(i, j int) bool {
+		return list2[i].UpkeepID < list2[j].UpkeepID
+	})
+	sort.Slice(list3, func(i, j int) bool {
+		return list3[i].UpkeepID < list3[j].UpkeepID
+	})
+	sort.Slice(list4, func(i, j int) bool {
+		return list4[i].UpkeepID < list4[j].UpkeepID
+	})
+
+	assert.NotEqual(t, list1, list2, "list1 vs list2")
+	assert.NotEqual(t, list1, list3, "list1 vs list3")
+	assert.NotEqual(t, list1, list4, "list1 vs list4")
 }
 
-func TestKeeperDB_EligibleUpkeeps_KeepersRotate(t *testing.T) {
+func TestKeeperDB_EligibleUpkeeps_SkipIfLastPerformedByCurrentKeeper(t *testing.T) {
```

Comment:
I couldn't find by test names, is there a test for the buddy logic?

Reply:
added one to show the buddy coverage